Applications/Spire/Source/Ui/InfoTip.cpp

```diff
@@ -0,0 +1,277 @@
+#include <QScreen>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/Box.hpp"
+
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  const auto DEFAULT_SHOW_DELAY_MS = 500;
+  const auto DROP_SHADOW_SIZE = 5;
+  const auto DROP_SHADOW_COLOR = QColor(0, 0, 0, 63);
+  const auto FADE_IN_MS = 100;
+
+  auto ARROW_SIZE() {
+    static auto size = scale(14, 7);
+    return size;
+  }
+
+  auto ARROW_X_POSITION() {
+    static auto x = scale_width(6);
+    return x;
+  }
+
+  auto DROP_SHADOW_HEIGHT() {
+    static auto height = scale_height(DROP_SHADOW_SIZE);
+    return height;
+  }
+
+  auto DROP_SHADOW_WIDTH() {
+    static auto width = scale_width(DROP_SHADOW_SIZE);
+    return width;
+  }
+
+  auto Y_OFFSET() {
+    static auto offset = scale_height(3);
+    return offset;
+  }
+
+  auto DEFAULT_STYLE() {
+    auto style = StyleSheet();
+    style.get(Any()).
+      set(BackgroundColor(QColor::fromRgb(0xFF, 0xFF, 0xFF))).
+      set(border_color(QColor::fromRgb(0xA0, 0xA0, 0xA0)));
+    return style;
+  }
+}
+
+InfoTip::InfoTip(QWidget* body, QWidget* parent)
+    : QWidget(parent, Qt::FramelessWindowHint | Qt::Tool |
+        Qt::NoDropShadowWindowHint | Qt::WindowDoesNotAcceptFocus),
+      m_body(body),
+      m_is_interactive(false) {
+  setAttribute(Qt::WA_ShowWithoutActivating);
+  setAttribute(Qt::WA_TranslucentBackground);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins(get_margins());
+  layout->addWidget(m_body);
+  m_show_timer.setInterval(DEFAULT_SHOW_DELAY_MS);
+  m_show_timer.setSingleShot(true);
+  connect(&m_show_timer, &QTimer::timeout, this, &InfoTip::on_show_timeout);
+  parent->installEventFilter(this);
+  connect_style_signal(*this, [=] { on_style(); });
+  set_style(*this, DEFAULT_STYLE());
+}
+
+bool InfoTip::eventFilter(QObject* watched, QEvent* event) {
+  switch(event->type()) {
+    case QEvent::Enter:
+      m_show_timer.start();
+      break;
+    case QEvent::MouseMove:
+      if(!parentWidget()->rect().contains(
+          static_cast<QMouseEvent*>(event)->pos())) {
+        fade_out();
+      }
+      break;
+    case QEvent::Leave:
+      if(!(m_is_interactive && hover_rect().contains(QCursor::pos()))) {
+        fade_out();
+      }
+      break;
+    case QEvent::WindowDeactivate:
+      fade_out();
+      break;
+    case QEvent::ToolTip:
+      return true;
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void InfoTip::leaveEvent(QEvent* event) {
+  fade_out();
+  QWidget::leaveEvent(event);
+}
+
+void InfoTip::paintEvent(QPaintEvent* event) {
+  auto painter = QPainter(this);
+  painter.drawPixmap(0, 0, render_background());
+}
+
+void InfoTip::set_interactive(bool is_interactive) {
+  m_is_interactive = is_interactive;
+}
+
+void InfoTip::fade_in() {
+  fade_window(this, false, FADE_IN_MS);
+}
+
+void InfoTip::fade_out() {
+  auto animation = fade_window(this, true, FADE_IN_MS);
+  connect(animation, &QPropertyAnimation::finished, [&] {
```

Comment:
Why is the lambda expression capturing by reference?

Reply:
I think I should capture 'this' by value, though.

Reply:
Capture by value when the lifetime of the closure evaluated by the lambda expression is longer than the lambda expression's lexical scope.